### PR TITLE
[MAINT] Refactor and clean our settings validation logs

### PR DESCRIPTION
**Pull request by ThomasHepworth**
### Type of PR

- [ ] BUG
- [ ] FEAT
- [x] MAINT
- [ ] DOC

### Is your Pull Request linked to an existing Issue or Pull Request?
Builds on https://github.com/moj-analytical-services/splink/pull/1522.

### Give a brief description for the solution you have provided
This PR refactors a lot of the existing code, in order to streamline the codebase and also makes fundamental changes to the logs to improve outputs.

The goals of this PR are:
* To make future development simpler by reducing the complexity of the code.
* Improve code readability.
* Improve docs readability.

<hr>

#### Main changes

**Refactoring**
* Pulled out `parse_columns_in_sql` so it can be more easily reused within the wider codebase.
* Refactored and cleaned up a lot of the existing code. If there's anything that's hard to follow, let me know as the aim here is to make the code much simpler to follow and expand, as and when needed.
* Pulled the existing test - `test_columns_from_settings` - apart, allowing us to individually verify each validation check is working as expected, rather than everything in one block. This grants us more flexibility and easier bug discovery.

**Log outputs**
* The logs now print in blocks, meaning the text output within user logging files is much easier to interpret. Previously, each paragraph was printed to the console individually, making the logs messier than they need to be.
* The logs are far cleaner. They also no longer depend on python's built in colours to help with readability, meaning they print properly across more systems and inside log text files.

**New log output**
The output from the settings logger looks now looks like so:
Screenshot 2023-10-05 at 12 08 36

And just for reference, the existing logs looked like so when printed inside a logging file:
Screenshot 2023-10-05 at 16 21 04

It's decipherable, but not ideal.

Documentation will follow in a future pull request.

### PR Checklist

- [ ] Added documentation for changes
- [ ] Added feature to example notebooks or tutorial (if appropriate)
- [x] Added tests (if appropriate)
- [x] Made changes based off the latest version of Splink
- [x] Run the [linter](https://moj-analytical-services.github.io/splink/dev_guides/changing_splink/lint.html)

### CODE REVIEW

**Comment by ThomasHepworth** [comment]
Let me know if anything is unclear, needs better documentation or if there's a better way to write something.

I generated the logs in the screenshot using the following code...
```py
from copy import deepcopy

import pandas as pd

from splink.duckdb.duckdb_linker import DuckDBLinker
from tests.basic_settings import get_settings_dict

df = pd.read_csv("./tests/datasets/fake_1000_from_splink_demos.csv")
settings = get_settings_dict()

# Create some wonky comparison levels
email_cc = {
    "output_column_name": "email",
    "comparison_levels": [
        {
            "sql_condition": "email_l IS NULL OR email_r IS NULL",
            "label_for_charts": "Comparison includes null",
            "is_null_level": True,
        },
        {
            "sql_condition": "levenshtein(emails_l, test.email_r) < 3",
            "label_for_charts": "Levenshtein < 3",
        },
        {
            "sql_condition": "date_diff('day', email_date_l, email_r)",
            "label_for_charts": "Random date diff",
        },
        {
            "sql_condition": "ELSE",
            "label_for_charts": "All other comparisons",
        },
    ],
}

city_cc = {
    "output_column_name": "city",
    "comparison_levels": [
        {
            "sql_condition": "city_l IS NULL OR city_r IS NULL",
            "label_for_charts": "Comparison includes null",
            "is_null_level": True,
        },
        {
            "sql_condition": "not_city_l = not_a_city_r",
            "label_for_charts": "Invalid exact SQL",
        },
        {
            "sql_condition": "city_test_l = city_r",
            "label_for_charts": "Exact match",
        },
        # Identical condition - should be ignored by the logger
        {
            "sql_condition": "city_test_l = city",
            "label_for_charts": "Exact match",
        },
        {
            # missing _l suffix
            "sql_condition": "sin(radians(\"city\"['lat']))",
            "label_for_charts": "Exact match",
        },
        {
            "sql_condition": "ELSE",
            "label_for_charts": "All other comparisons",
        },
    ],
}

settings["comparisons"][3:] = [email_cc, city_cc]
settings["blocking_rules_to_generate_predictions"] = [
    "i_dont_exist = r.i_do_exist",
    "l.round_two_check = r.email"
]
settings["unique_id_column_name"] = "mythical"

linker = DuckDBLinker(df, settings, validate_settings=True)
```

**Comment by ThomasHepworth** [comment]
Some of the tests seem to be a tad shaky. Also, some functions don't parse as expected due to a bug in SQLglot. But, nevertheless, this is getting there.

**Review by RossKen — APPROVED** [review]
Apologies for the delay @ThomasHepworth!

The restructure looks sensible and I agree, it will make extending more straightforward. The new blocks in the output are also a great improvement 🙌

In terms of the SQL glot issue, I don't think there is much we can do at this stage so I am happy for this to go in alongside your documentation PR

**Comment by ThomasHepworth** [comment]
https://github.com/moj-analytical-services/splink/pull/1636#pullrequestreview-1684795201

On the SQLglot problem - now that we've upgraded sqlglot, I believe the issue has gone away. I'll continue to monitor it, though.
